### ISSUE TITLE
Any chance we can get UPSERT in Java?

### ISSUE THREAD

**Issue by Queatz** [arangodb/arangodb-java-driver#167]
In my scenario, I don't always know if my objects have been stored yet.  Would be ideal to have i.e. `ArangoCollection.upsertDocument` and `ArangoCollection.upsertDocuments`

**Comment by mpv1989** [comment]
@Queatz Implementing a REST API for upsert is on our todo list. As soon as we implemented it on server side the driver will be updated too.

It's not that high prioritized so I don't know when it will be implemented. I'll keep you informed.

**Comment by Queatz** [comment]
@mpv1989 Sounds like a plan. Is there any way I can "watch" or follow that feature?

**Comment by mpv1989** [comment]
@Queatz currently it is only in our backlog, as soon as we delegate this task we will create an issue in our main repo. I will then link it here.

**Comment by mpv1989** [comment]
@Queatz the Java driver now support REPSERT (replace-insert) over
`ArangoCollection.insertDocument(doc, new DocumentCreateOptions().overwrite(true)`

**Comment by pvgoran** [comment]
@mpv1989 Replace-insert is not quite the same as Upsert.

Can we get this re-opened?

**Comment by dothebart** [comment]
This could be done after
https://github.com/arangodb/arangodb/issues/2542
is implemented.

**Comment by rashtao** [comment]
Already implemented here: https://github.com/arangodb/arangodb-java-driver/commit/0b4f4044a92f7ca5ba2d04c74a62147ed60bee80
and for the moment released in the 3.7 preview branch https://github.com/arangodb/arangodb-java-driver/releases/tag/v6.7.0_PREVIEW_3.7.0-alpha.2_0

**Comment by pvgoran** [comment]
@rashtao Looks interesting.

However, it doesn't allow to specify what attributes to look at when matching documents to update (similar to the `upsert` AQL instruction).

So if I have a document `{ "myid": 4, "unimportant": 720 }`, I can't update it with `{ "myid": 4, "unimportant": 910 }`, can I?

**Comment by rashtao** [comment]
@pvgoran The driver implementation is matching the updated API for creating documents (coming in ArangoDB 3.7), which will support an additional `overwriteMode` parameter:
https://www.arangodb.com/docs/devel/http/document-working-with-documents.html#create-document

For more customized control, you can run AQL queries from the driver.

**Comment by pvgoran** [comment]
@rashtao Hm, I might have misunderstood the tests before. Judging from the new documentation, the "replace/update" functionality is only triggered by matching the `_key` parameter, is this right? No way to match by anything else besides `_key`?

**Comment by rashtao** [comment]
@pvgoran yes, it is related to the document matched by the `_key` attrubute.

**Comment by pvgoran** [comment]
> @pvgoran yes, it is related to the document matched by the `_key` attrubute.

Then, this functionality is only tangentially related to the real UPSERT (well, at least to UPSERT as it's implemented in AQL).

**Comment by rashtao** [comment]
@pvgoran The driver responsibility is limited to call the server API and expose such operations as a Java API. More advanced operations can either be performed using AQL or implemented in an higher level libraries.
You might be interested in further discussing server API for creating documents in this thread: https://github.com/arangodb/arangodb/issues/2542

**Comment by pvgoran** [comment]
> The driver responsibility is limited to call the server API and expose such operations as a Java API.

Yes, I understand this. I'm only saying that this new functionality shouldn't be presented as a solution to the UPSERT request.

**Comment by rashtao** [comment]
Closing since the Java driver already implements the related server API for updating and replacing documents having the same keys. Further extension of server API can be discussed in https://github.com/arangodb/arangodb/issues/2542
